### PR TITLE
Add script for automatic generation of our documentation

### PR DESCRIPTION
**Pull request by AndreasBurger**

### CODE REVIEW

File: .ci/generate_documentation [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+context = ctx()
+config_factory = context.cfg_factory()
+gh_config = config_factory._cfg_element(cfg_type_name='github', cfg_name='github_com')
+
+with tempfile.TemporaryDirectory() as tmp_dir:
+    gitutil.clone_repository(
+        to_path=tmp_dir,
+        github_cfg=gh_config,
+        github_repo_path=github_owner_and_name,
```
Comment (ccwienk) [inline .ci/generate_documentation]: how about adding a new _optional_ argument "branch" to clone_repository (similar as git-clone offers it)? Sane default value might be '[refs/heads]/master'

File: .ci/generate_documentation [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    ctx,
+    fail,
+    which,
+)
+
+
+sphinx_executable = which('sphinx-build')
+
+source_repo_dir = pathlib.Path(check_env('SOURCE_PATH')).resolve()
+source_dir = check_env('SOURCE_DIRECTORY')
```
Comment (ccwienk) [inline .ci/generate_documentation]: please rename. how about [REL_?]DOC_ROOT_DIR

File: .ci/generate_documentation [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+github_owner_and_name = check_env('SOURCE_GITHUB_REPO_OWNER_AND_NAME')
+_, repo_name = github_owner_and_name.split('/')
+
+context = ctx()
+config_factory = context.cfg_factory()
+gh_config = config_factory._cfg_element(cfg_type_name='github', cfg_name='github_com')
```
Comment (ccwienk) [inline .ci/generate_documentation]: use github/util.py's
```
@functools.lru_cache()
418 def github_cfg_for_hostname(cfg_factory, host_name):
```

File: .ci/generate_documentation [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    )
+
+    tmp_repo_dir = pathlib.Path(tmp_dir).resolve() / repo_name
+
+    git.cmd.Git(tmp_repo_dir).checkout(['-f','gh-pages'])
+
+    git_helper = gitutil.GitHelper(
+        repo=tmp_repo_dir,
+        github_cfg=gh_config,
+        github_repo_path=github_owner_and_name,
```
Comment (ccwienk) [inline .ci/generate_documentation]: please rename github_owner_and_name to "target_repo_path" or sth similar (or maybe "output_repo_path")? Some name that indicates the semantics that this script will push to it

File: .ci/generate_documentation [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    # Run sphinx-build with given source and target dirs
+    p = subprocess.Popen(
+        [
+            sphinx_executable,
+            '-b', 'html',
+            '-a',
+            input_path,
+            output_path
+        ],
+        stdout=subprocess.PIPE,
```
Comment (ccwienk) [inline .ci/generate_documentation]: rm

File: .ci/generate_documentation [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        stderr=subprocess.PIPE,
```
Comment (ccwienk) [inline .ci/generate_documentation]: rm

File: .ci/generate_documentation [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        repo=tmp_repo_dir,
+        github_cfg=gh_config,
+        github_repo_path=github_owner_and_name,
+    )
+
+    input_path = source_repo_dir / source_dir
+    output_path = tmp_repo_dir
+
+    # Run sphinx-build with given source and target dirs
+    p = subprocess.Popen(
```
Comment (ccwienk) [inline .ci/generate_documentation]: subprocess.run?

File: .ci/generate_documentation [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        [
+            sphinx_executable,
+            '-b', 'html',
+            '-a',
+            input_path,
+            output_path
+        ],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        universal_newlines=True,
```
Comment (ccwienk) [inline .ci/generate_documentation]: rm

File: .ci/generate_documentation [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    )
+    stdout, stderr = p.communicate()
```
Comment (ccwienk) [inline .ci/generate_documentation]: replace with subprocess.call

File: .ci/generate_documentation [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    return_code = p.returncode
+    print(stdout)
+    print(stderr)
+
+    if p.returncode != 0:
```
Comment (ccwienk) [inline .ci/generate_documentation]: adjust

File: .ci/generate_documentation [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+# Verify that required env_vars are set
+sphinx_executable = which('sphinx-build')
+source_repo_dir = pathlib.Path(check_env('SOURCE_PATH')).resolve()
+source_relative_doc_root_dir = check_env('REL_DOC_ROOT_DIR')
+
+# Verify optional env_vars
+# In anticipation of changes to pipeline contract, use '<logical_name>_COMPONENT_NAME>
+# (e.g.: 'github.com/gardener/cc-utils')
+# TODO: rm comment once changes are made
+target_component_name = os.environ.get('TARGET_COMPONENT_NAME', None)
```
Comment (ccwienk) [inline .ci/generate_documentation]: immediately make this a hard requirement (util.check_env). use step: <step_name>.vars to pass env vars

File: .ci/generate_documentation [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    # Run sphinx-build with given source and target dirs
+    subprocess.run(
+        [
+            sphinx_executable,
+            '-b', 'html',
+            '-a',
+            input_path,
+            output_path
+        ],
+        check=True,
```
Comment (ccwienk) [inline .ci/generate_documentation]: how about check_run?

**Comment by AndreasBurger** [comment]
Closed; As mentioned during the review the resulting script is included in our pipelines-repository

**Pull request closed without merging**